kani-compiler/src/kani_middle/provide.rs

```diff
@@ -55,7 +55,9 @@ fn run_kani_mir_passes<'tcx>(
     body: &'tcx Body<'tcx>,
 ) -> &'tcx Body<'tcx> {
     tracing::debug!(?def_id, "Run Kani transformation passes");
-    stubbing::transform(tcx, def_id, body)
+    let mut transformed_body = stubbing::transform(tcx, def_id, body);
+    stubbing::transform_extern_functions(tcx, &mut transformed_body);
+    tcx.arena.alloc(transformed_body)
```

Comment:
Hmm, I'm not sure we need this allocation every time. Both `transform` and `transform_extern_function` could return a `Cow`. Only if at the end we have a `Cow::Owned` we would need to allocate a new body.

What I'm imagining is something like this

```rs
let transformed_body = /* in stubbing::transform */ |old_body| {
  if let Some(replacement) {
    Cow::Owned(/* replacement transformed */)
  } else {
    Cow::Borrowed(old_body)
  };
  let extern_transformed = /* in transform_extern_functions */ {
    let mut body = transformed_body.clone(); // We a local copy to mutate
    let body_was_borrowed = transformed_body.is_borrowed();
    vis.visit_body(&mut body);
    if body == old_body.borrow() {
      // No change happened, we can just use the old body
      transformed_body
    } else {
      Cow::Owned(body)
    };
    match extern_transformed {
      Cow::Owned(new_body) => tcx.arena.alloc(body),
      Cow::Borrowed(old_body) => old_body,
    }
```

Hope this makes sense.

This way we avoid the additional allocation in the arena which otherwise would happen for every single function.

Also it means that the body handed to `transform_extern_functions` is used borrowed. That means while it still needs to make a copy to mutate you can elide the copy you currently store in the visitor and just store a `&LocalDecls` in there instead which is all you need for `operand.ty` actually.

Reply:
We could do that, but wouldn't it be easier to do that outside of the transformation. I.e.: 
```rust
If *orig_body != body {
    tcx.arena.alloc(body)
} else {
    old.body()
}
```

Reply:
Oh yeah I meant outside of the transformation. I just inlined the code to show the pipeline.
I have to say though your method is a lot simpler 😂

Reply:
It was a team effort. :)

Reply:
Body is not comparable, so we would either have to implement our own equality, use a `StableHash`, or implement the solution using the COW pattern as suggested by @JustusAdam. Since this change is somewhat orthogonal to this PR, I'm gonna track on this separate issue https://github.com/model-checking/kani/issues/2664.